File: Python/loo1-python/Exemplo3.py
import sys
import os
sys.path.insert(0, os.path.abspath(os.path.dirname(__file__)))

from Programa import Programa
from command.atribuicao import Atribuicao
from command.chamada_metodo import ChamadaMetodo
from command.comando_declaracao import ComandoDeclaracao
from command.new import New
from command.sequencia_comando import SequenciaComando
from command.write import Write
from command.skip import Skip
from command.if_then_else import IfThenElse
from declaration.dec_classe_simples import DecClasseSimples
from declaration.dec_procedimento_composta import DecProcedimentoComposta
from declaration.dec_procedimento_simples import DecProcedimentoSimples
from declaration.dec_variavel_objeto import DecVariavelObjeto
from declaration.declaracao_variavel import DeclaracaoVariavel
from declaration.declaracao_composta import DeclaracaoComposta
from declaration.lista_declaracao_parametro import ListaDeclaracaoParametro
from declaration.dec_parametro import DecParametro
from expression.acesso_atributo_this import AcessoAtributoThis
from expression.exp_equals import ExpEquals
from expression.exp_not import ExpNot
from expression.id import Id
from expression.this_expressao import This
from expression.valor_inteiro import ValorInteiro
from expression.valor_null import ValorNull
from memory.contexto_compilacao_oo1 import ContextoCompilacaoOO1
from memory.contexto_execucao_oo1 import ContextoExecucaoOO1
from memory.lista_valor import ListaValor
from util.tipo_classe import TipoClasse
from util.tipo_primitivo import TipoPrimitivo
from expression.lista_expressao import ListaExpressao

def executar_exemplo(prog, descricao, ambiente):
    print("-" * 50)
    print("Programa: \n" + descricao + "\n")
    try:
        bem_tipado = prog.checaTipo(ContextoCompilacaoOO1(ListaValor()))
        print("Bem tipado:", bem_tipado)
        if bem_tipado:
            saida = prog.executar(ambiente)
            print("Saida:", end=" ")
            while saida is not None and saida.length() > 0:
                print(saida.getHead(), end=" ")
                saida = saida.getTail()
            print()
        else:
            print("Erro de tipo: execucao abortada.")
    except Exception as e:
        import traceback
        print("Excecao capturada: " + str(e))
        traceback.print_exc()
    print("-" * 50)

if __name__ == "__main__":
    idLValor = Id("LValor")
    tipoLValor = TipoClasse(idLValor)

    decValor = DeclaracaoVariavel(TipoPrimitivo.INTEIRO, Id("valor"), ValorInteiro(-100))
    decProx = DeclaracaoVariavel(tipoLValor, Id("prox"), ValorNull())
    decVariaveis = DeclaracaoComposta(decValor, decProx)

    idV = Id("v")
    paramV = ListaDeclaracaoParametro(DecParametro(idV, TipoPrimitivo.INTEIRO))

    condInsere = ExpEquals(AcessoAtributoThis(This(), Id("valor")), ValorInteiro(-100))
    thisValorV = Atribuicao(AcessoAtributoThis(This(), Id("valor")), idV)
    thisProxNew = New(AcessoAtributoThis(This(), Id("prox")), idLValor)
    thenInsere = SequenciaComando(thisValorV, thisProxNew)

    argV = ListaExpressao(idV)
    elseInsere = ChamadaMetodo(AcessoAtributoThis(This(), Id("prox")), Id("insere"), argV)

    ifInsere = IfThenElse(condInsere, thenInsere, elseInsere)
    procInsere = DecProcedimentoSimples(Id("insere"), paramV, ifInsere)

    writeValor = Write(AcessoAtributoThis(This(), Id("valor")))
    proxNull = ExpEquals(AcessoAtributoThis(This(), Id("prox")), ValorNull())
    condPrint = ExpNot(proxNull)

    thenPrint = ChamadaMetodo(AcessoAtributoThis(This(), Id("prox")), Id("print"), ListaExpressao())
    elsePrint = Skip()

    ifPrint = IfThenElse(condPrint, thenPrint, elsePrint)
    corpoPrint = SequenciaComando(writeValor, ifPrint)

    procPrint = DecProcedimentoSimples(Id("print"), ListaDeclaracaoParametro(), corpoPrint)
    decProcedimentos = DecProcedimentoComposta(procInsere, procPrint)

    decClasse = DecClasseSimples(idLValor, decVariaveis, decProcedimentos)

    decLv = DecVariavelObjeto(tipoLValor, Id("lv"), idLValor)

    arg3 = ListaExpressao(ValorInteiro(3))
    lvInsere3 = ChamadaMetodo(Id("lv"), Id("insere"), arg3)

    arg4 = ListaExpressao(ValorInteiro(4))
    lvInsere4 = ChamadaMetodo(Id("lv"), Id("insere"), arg4)

    lvPrint = ChamadaMetodo(Id("lv"), Id("print"), ListaExpressao())

    seqPrin1 = SequenciaComando(lvInsere3, lvInsere4)
    seqPrin2 = SequenciaComando(seqPrin1, lvPrint)

    comDeclaracao = ComandoDeclaracao(decLv, seqPrin2)
    prog = Programa(decClasse, comDeclaracao)

    executar_exemplo(prog, "LValor lv := new LValor;\nlv.insere(3);\nlv.insere(4);\nlv.print()", ContextoExecucaoOO1())

File: Java/loo1/plp/orientadaObjetos1/Exemplo3.java
package plp.orientadaObjetos1;

import plp.orientadaObjetos1.comando.*;
import plp.orientadaObjetos1.declaracao.classe.*;
import plp.orientadaObjetos1.declaracao.procedimento.*;
import plp.orientadaObjetos1.declaracao.variavel.*;
import plp.orientadaObjetos1.expressao.*;
import plp.orientadaObjetos1.expressao.binaria.*;
import plp.orientadaObjetos1.expressao.unaria.*;
import plp.orientadaObjetos1.expressao.leftExpression.*;
import plp.orientadaObjetos1.expressao.valor.*;
import plp.orientadaObjetos1.memoria.*;
import plp.orientadaObjetos1.memoria.colecao.ListaValor;
import plp.orientadaObjetos1.util.*;

public class Exemplo3 {

    public static void executar(Programa prog, String descricao, ContextoExecucaoOO1 ambiente) {
        System.out.println("--------------------------------------------------");
        System.out.println("Programa: \n" + descricao + "\n");
        try {
            boolean bemTipado = prog.checaTipo(new ContextoCompilacaoOO1(new ListaValor()));
            System.out.println("Bem tipado: " + bemTipado);
            if (bemTipado) {
                ListaValor saida = prog.executar(ambiente);
                System.out.print("Saida: ");
                while (saida.length() > 0) {
                    System.out.print(saida.getHead() + " ");
                    saida = (ListaValor) saida.getTail();
                }
                System.out.println();
            } else {
                System.out.println("Erro de tipo: execucao abortada.");
            }
        } catch (Exception e) {
            System.out.println("Excecao capturada: " + e.getMessage());
            e.printStackTrace();
        }
        System.out.println("--------------------------------------------------");
    }

    public static void main(String[] args) {
        try {
            // Id LValor
            Id idLValor = new Id("LValor");
            Tipo tipoLValor = new TipoClasse(idLValor);

            // int valor = -100
            DecVariavel decValor = new SimplesDecVariavel(
                new TipoPrimitivo(TipoPrimitivo.INTEIRO), 
                new Id("valor"), 
                new ValorInteiro(-100)
            );

            // LValor prox = null
            DecVariavel decProx = new SimplesDecVariavel(
                tipoLValor, 
                new Id("prox"), 
                new ValorNull()
            );

            // CompostaDecVariavel(decValor, decProx)
            DecVariavel decVariaveis = new CompostaDecVariavel(decValor, decProx);

            // proc insere(int v) { ... }
            Id idV = new Id("v");
            ListaDeclaracaoParametro paramV = new ListaDeclaracaoParametro(
                new DecParametro(idV, new TipoPrimitivo(TipoPrimitivo.INTEIRO))
            );

            // if ((this).valor == -100)
            Expressao condInsere = new ExpEquals(
                new AcessoAtributoThis(new This(), new Id("valor")), 
                new ValorInteiro(-100)
            );

            // then { this.valor := v; this.prox := new LValor }
            Comando thisValorV = new Atribuicao(new AcessoAtributoThis(new This(), new Id("valor")), idV);
            Comando thisProxNew = new New(new AcessoAtributoThis(new This(), new Id("prox")), idLValor);
            Comando thenInsere = new Sequencial(thisValorV, thisProxNew);

            // else { (this).prox.insere(v) }
            ListaExpressao argV = new ListaExpressao(idV);
            Comando elseInsere = new ChamadaMetodo(new AcessoAtributoThis(new This(), new Id("prox")), new Id("insere"), argV);

            Comando ifInsere = new IfThenElse(condInsere, thenInsere, elseInsere);
            DecProcedimento procInsere = new DecProcedimentoSimples(new Id("insere"), paramV, ifInsere);

            // proc print() { ... }
            Comando writeValor = new Write(new AcessoAtributoThis(new This(), new Id("valor")));

            // if (not(this.prox == null))
            Expressao proxNull = new ExpEquals(new AcessoAtributoThis(new This(), new Id("prox")), new ValorNull());
            Expressao condPrint = new ExpNot(proxNull);

            // then { (this).prox.print() }
            Comando thenPrint = new ChamadaMetodo(new AcessoAtributoThis(new This(), new Id("prox")), new Id("print"), new ListaExpressao());

            // else { skip }
            Comando elsePrint = new Skip();

            Comando ifPrint = new IfThenElse(condPrint, thenPrint, elsePrint);
            Comando corpoPrint = new Sequencial(writeValor, ifPrint);

            DecProcedimento procPrint = new DecProcedimentoSimples(new Id("print"), new ListaDeclaracaoParametro(), corpoPrint);

            DecProcedimento decProcedimentos = new DecProcedimentoComposta(procInsere, procPrint);

            DecClasse decClasse = new DecClasseSimples(idLValor, decVariaveis, decProcedimentos);

            // Bloco Principal
            // LValor lv := new LValor;
            DecVariavelObjeto decLv = new DecVariavelObjeto(tipoLValor, new Id("lv"), idLValor);

            // lv.insere(3);
            ListaExpressao arg3 = new ListaExpressao(new ValorInteiro(3));
            Comando lvInsere3 = new ChamadaMetodo(new Id("lv"), new Id("insere"), arg3);

            // lv.insere(4);
            ListaExpressao arg4 = new ListaExpressao(new ValorInteiro(4));
            Comando lvInsere4 = new ChamadaMetodo(new Id("lv"), new Id("insere"), arg4);

            // lv.print();
            Comando lvPrint = new ChamadaMetodo(new Id("lv"), new Id("print"), new ListaExpressao());

            // Sequencia principal
            Comando seqPrin1 = new Sequencial(lvInsere3, lvInsere4);
            Comando seqPrin2 = new Sequencial(seqPrin1, lvPrint);

            Comando comDeclaracao = new ComDeclaracao(decLv, seqPrin2);
            Programa prog = new Programa(decClasse, comDeclaracao);

            // Executar
            executar(prog, "LValor lv := new LValor;\nlv.insere(3);\nlv.insere(4);\nlv.print()", new ContextoExecucaoOO1());
            
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
